wasm-node/src/node.rs

```diff
@@ -0,0 +1,179 @@
+        info!(
+            "Initialised store with head height: {:?}",
+            store.get_head_height()
+        );
+
+        let node = Node::new(NodeConfig {
+            network_id: network_id.to_string(),
+            genesis_hash: config.genesis_hash,
+            p2p_local_keypair: config.p2p_local_keypair,
+            p2p_bootnodes: config.p2p_bootnodes,
+            p2p_listen_on: vec![],
+            store,
+        })
+        .await
+        .context("Failed to start node")
+        .unwrap_throw();
+
+        Self { node }
+    }
+
+    pub fn local_peer_id(&self) -> String {
+        self.node.p2p().local_peer_id().to_string()
+    }
+
+    pub async fn wait_connected(&self) {
+        self.node.p2p().wait_connected().await.unwrap_throw();
+    }
+
+    pub async fn wait_connected_trusted(&self) {
+        self.node
+            .p2p()
+            .wait_connected_trusted()
+            .await
+            .unwrap_throw();
+    }
+
+    pub async fn network_info(&self) -> JsValue {
+        // NetworkInfo isn't serializable
+        //let info = self.node.p2p().network_info().await.unwrap_throw();
+        //to_value(&info).unwrap_throw()
+        unimplemented!()
+    }
+
+    pub async fn get_head_header(&self) -> JsValue {
+        let eh = self.node.p2p().get_head_header().await.unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_header(&self, hash: &str /*&Hash*/) -> JsValue {
+        let hash = hash.parse().unwrap_throw();
+        let eh = self.node.p2p().get_header(hash).await.unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_header_by_height(&self, height: u64) -> JsValue {
+        let eh = self
+            .node
+            .p2p()
+            .get_header_by_height(height)
+            .await
+            .unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_verified_headers_range() {
+        unimplemented!()
+    }
+
+    pub async fn listeners(&self) -> Array {
+        self.node
+            .p2p()
+            .listeners()
+            .await
+            .unwrap_throw()
+            .iter()
+            .map(ToString::to_string)
+            .map(JsValue::from)
+            .collect::<Array>()
+    }
+
+    pub async fn connected_peers(&self) -> Array {
+        self.node
+            .p2p()
+            .connected_peers()
+            .await
+            .unwrap_throw()
+            .iter()
+            .map(ToString::to_string)
+            .map(JsValue::from)
+            .collect::<Array>()
+    }
+
+    pub async fn syncer_info(&self) -> JsValue {
+        let syncer_info = self.node.syncer().info().await.unwrap_throw();
+        to_value(&syncer_info).unwrap_throw()
+    }
+}
+
+#[wasm_bindgen]
+impl WasmNodeConfig {
+    #[wasm_bindgen(constructor)]
+    pub fn new(network: Network, genesis_hash: JsString, bootnodes: Vec<JsString>) -> Self {
```

Comment:
I feel these are not needed since we have setters. What is the idiomatic way in JS?

Reply:
this is for the `new WasmNodeConfig(...)` syntax
I'm not sure what the difference is with wasm bindgen, but in javascript it will have a separation of object and instance, eg. the methods will be stored once in the object and shared by instances, if you would just return objects without new then they would allocate a separated function each

Reply:
Constructor here will add, as I understand it, JS's idomatice `new ...` way to create class instance, as @zvolin mentioned.
I've added getters and setters so that js code can interact with config like a normal js object. Most of the fields aren't directly representable in js, so I had to `#[wasm_bindgen(skip)]` them, which resulted in opaque class without members.